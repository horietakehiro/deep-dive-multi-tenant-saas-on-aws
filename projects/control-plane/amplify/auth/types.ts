/**
 * サインアップ時に必要なカスタム属性名
 */
export const SIGNUP_CUSTOM_USER_ATTRIBUTES = {
  TENANT_ID: "custom:tenantId",
  TENANT_NAME: "custom:tenantName",
} as const;
/**
 * サインアップ時に必要なカスタム属性
 */
export type SignUpUserAttributes = {
  [key in
    | (typeof SIGNUP_CUSTOM_USER_ATTRIBUTES)[keyof typeof SIGNUP_CUSTOM_USER_ATTRIBUTES]
    | "email"]: string;
};
