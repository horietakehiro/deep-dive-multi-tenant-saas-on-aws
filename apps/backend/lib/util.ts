import { NotImplementedError } from "./domain/model/error";

export const notImplementedFn = () => {
  throw NotImplementedError;
};
