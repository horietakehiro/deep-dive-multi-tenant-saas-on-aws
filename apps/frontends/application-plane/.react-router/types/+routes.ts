// Generated by React Router

import "react-router"

declare module "react-router" {
  interface Register {
    pages: Pages
    routeFiles: RouteFiles
  }
}

type Pages = {
  "/": {
    params: {};
  };
  "/appointments": {
    params: {};
  };
};

type RouteFiles = {
  "root.tsx": {
    id: "root";
    page: "/" | "/appointments";
  };
  "./layouts/auth.tsx": {
    id: "layouts/auth";
    page: "/" | "/appointments";
  };
  "./layouts/dashboard.tsx": {
    id: "layouts/dashboard";
    page: "/" | "/appointments";
  };
  "./routes/home.tsx": {
    id: "routes/home";
    page: "/";
  };
  "./routes/appointments.tsx": {
    id: "routes/appointments";
    page: "/appointments";
  };
};