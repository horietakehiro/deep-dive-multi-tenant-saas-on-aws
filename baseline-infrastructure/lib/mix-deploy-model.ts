import * as fs from "fs";
import * as cdk from "aws-cdk-lib";
import { Construct } from "constructs";
import {
  aws_amplify as amplify,
  aws_iam as iam,
  aws_certificatemanager as acm,
  aws_route53 as route53,
} from "aws-cdk-lib";
import * as path from "node:path";
export interface PlaneProps {
  certArn: string;
  appName: "control-plane" | "intersection";
}
class Plane extends Construct {
  constructor(scope: Construct, id: string, props: PlaneProps) {
    super(scope, id);
    const accessToken = new cdk.CfnParameter(this, "AccessToken", {
      noEcho: true,
      type: "String",
    });
    const serviceRole = new iam.Role(this, "ServiceRole", {
      assumedBy: new iam.ServicePrincipal("amplify.amazonaws.com"),
      managedPolicies: [
        iam.ManagedPolicy.fromAwsManagedPolicyName(
          "service-role/AmplifyBackendDeployFullAccess"
        ),
        iam.ManagedPolicy.fromAwsManagedPolicyName("ReadOnlyAccess"),
      ],
      inlinePolicies: {
        root: new iam.PolicyDocument({
          statements: [
            new iam.PolicyStatement({
              effect: iam.Effect.ALLOW,
              actions: [
                "logs:CreateLogStream",
                "logs:PutLogEvents",
                "logs:CreateLogGroup",
                "logs:DescribeLogGroups",
              ],
              resources: ["*"],
            }),
          ],
        }),
      },
    });
    const app = new amplify.CfnApp(this, "App", {
      name: props.appName,
      accessToken: cdk.Fn.ref(accessToken.logicalId),
      autoBranchCreationConfig: {
        enableAutoBranchCreation: false,
      },
      basicAuthConfig: {
        enableBasicAuth: false,
      },
      buildSpec: fs
        .readFileSync(path.join(__dirname, "..", "amplify.yml"))
        .toString("utf-8"),
      cacheConfig: {
        type: "AMPLIFY_MANAGED_NO_COOKIES",
      },
      computeRoleArn: undefined,
      customHeaders: undefined,
      customRules: [
        {
          source: "/<*>",
          target: "/index.html",
          status: "404-200",
        },
      ],
      description: "control plane",
      enableBranchAutoDeletion: false,
      environmentVariables: [
        {
          name: "AMPLIFY_DIFF_DEPLOY",
          value: "false",
        },
        {
          name: "AMPLIFY_MONOREPO_APP_ROOT",
          value: `projects/${props.appName}`,
        },
      ],
      iamServiceRole: serviceRole.roleArn,
      oauthToken: undefined,
      platform: "WEB",
      repository:
        "https://github.com/horietakehiro/deep-dive-multi-tenant-saas-on-aws",
    });

    new amplify.CfnBranch(this, "MainBranch", {
      appId: app.attrAppId,
      branchName: "main",
      stage: "PRODUCTION",
      enableAutoBuild: true,
      framework: "web",
      // backend
    });
    new amplify.CfnDomain(this, "Domain", {
      appId: app.attrAppId,
      domainName: "ht-burdock.com",
      subDomainSettings: [
        {
          branchName: "main",
          prefix: props.appName,
        },
      ],
      autoSubDomainCreationPatterns: [],
      certificateSettings: {
        certificateType: "CUSTOM",
        customCertificateArn: props.certArn,
      },
    });
  }
}

export interface MixDeployModelStackProps extends cdk.StackProps {
  cert: acm.ICertificate;
}
export class MixDeployModelStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props: MixDeployModelStackProps) {
    super(scope, id, props);

    new Plane(this, "ControlPlane", {
      certArn: props.cert.certificateArn,
      appName: "control-plane",
    });
    new Plane(this, "ApplicationPlane", {
      certArn: props.cert.certificateArn,
      appName: "intersection",
    });
  }
}

export interface CustomCertStackProps extends cdk.StackProps {
  domaineName: string;
  hostedZoneId: string;
}
export class CustomCertStack extends cdk.Stack {
  public readonly certificate: acm.ICertificate;
  constructor(scope: Construct, id: string, props: CustomCertStackProps) {
    super(scope, id, props);
    this.certificate = new acm.Certificate(this, "Certificate", {
      domainName: props.domaineName,
      validation: acm.CertificateValidation.fromDns(
        route53.PublicHostedZone.fromHostedZoneId(
          this,
          "HostedZone",
          props.hostedZoneId
        )
      ),
    });
  }
}
