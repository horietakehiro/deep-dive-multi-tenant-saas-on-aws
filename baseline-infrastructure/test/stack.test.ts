import { App, Stack } from "aws-cdk-lib";
import { ControlPlane } from "../lib/full-stack-silo-deploy-model";
import { AWS_AMPLIFY_APP } from "@horietakehiro/aws-cdk-utul/lib/types/cfn-resource-types";
import {
  TypedTemplate,
  ExtraMatch,
} from "@horietakehiro/aws-cdk-utul/lib/assertions";

describe("コントロールプレーンコンストラクタ", () => {
  const app = new App();
  const stack = new Stack(app, "Test");
  new ControlPlane(stack, "ControlPlane", {
    branch: "main",
    certArn: "arn:aws:acm:Region:444455556666:certificate/certificate_ID",
  });
  const template = TypedTemplate.fromStack(stack);

  test("セキュア文字列パラメータとしてGitHubアクセストークンをAmplifyアプリケーションに設定している", () => {
    const params = template.findParameters("*", {
      NoEcho: true,
      Type: "String",
    });
    expect(params.length).toBe(1);
    const { id } = params[0];
    template.hasResource(
      AWS_AMPLIFY_APP({
        Properties: {
          AccessToken: ExtraMatch.ref(id),
        },
      })
    );
  });
  test("スナップショットテスト", () => {
    expect(template.template).toMatchSnapshot();
  });
});
