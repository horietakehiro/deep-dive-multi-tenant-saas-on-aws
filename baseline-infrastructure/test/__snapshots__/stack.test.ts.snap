// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`コントロールプレーンコンストラクタ > スナップショットテスト 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
    "ControlPlaneAccessTokenF660E7BE": {
      "NoEcho": true,
      "Type": "String",
    },
  },
  "Resources": {
    "ControlPlaneApp8CAC7211": {
      "Properties": {
        "AccessToken": {
          "Ref": "ControlPlaneAccessTokenF660E7BE",
        },
        "AutoBranchCreationConfig": {
          "EnableAutoBranchCreation": false,
        },
        "BasicAuthConfig": {
          "EnableBasicAuth": false,
        },
        "BuildSpec": "applications:
  - appRoot: apps/control-plane
    env:
      variables:
        W: apps/control-plane
    frontend:
      buildPath: '/'
      phases:
        preBuild:
          commands:
            - npm install -w \${W}
        build:
          commands:
            - npm run build -w \${W}
      artifacts:
        baseDirectory: apps/control-plane/build/client
        files:
          - "**/*"
      cache:
        paths:
          - .npm/**/*
          - node_modules/**/*
    backend:
      buildPath: "/"
      phases:
        build:
          commands:
            - npm ci --cache .npm --prefer-offlinei -w \${W}
            - npm install --save-dev @aws-amplify/backend @aws-amplify/backend-cli --cache .npm --prefer-offlinei
            - npm run -w \${W} ampx -- pipeline-deploy --branch $AWS_BRANCH --app-id $AWS_APP_ID
  - appRoot: apps/application-plane
    env:
      variables:
        W: apps/application-plane
        # CONTROL_PLANE_APP_ID: "アプリケーション作成時に設定される"
    frontend:
      buildPath: "/"
      phases:
        preBuild:
          commands:
            - npm install -w \${W}
        build:
          commands:
            - npm run build -w \${W}
      artifacts:
        baseDirectory: apps/application-plane/build/client
        files:
          - "**/*"
      cache:
        paths:
          - .npm/**/*
          - node_modules/**/*
    backend:
      phases:
        build:
          commands:

            - npm ci --cache .npm --prefer-offlinei -w \${W}
            - npm install --save-dev @aws-amplify/backend @aws-amplify/backend-cli --cache .npm --prefer-offlinei
            - sudo dnf install -y jq
            # # リソースを共有したい、作成済みのコントロールプレーン側のリソース設定をエクスポートする
            # コントロールプレーンのバックエンドスタックが更新中の場合はoutputsファイルを生成できないため、更新が完了するまで待機する
            - STACK_ARN=$(aws amplify get-branch --app-id  \${CONTROL_PLANE_APP_ID}  --branch-name \${AWS_BRANCH} | jq -r .branch.backend.stackArn)
            - STACK_STATUS=$(aws cloudformation describe-stacks --stack-name $STACK_ARN | jq -r .Stacks[0].StackStatus)
            - echo stack status of \${STACK_ARN} is \${STACK_STATUS}
            - | 
              while [[ $STACK_STATUS != *"COMPLETE" ]]
              do
                echo wait for 15 sec...
                sleep 15
                STACK_STATUS=$(aws cloudformation describe-stacks --stack-name $STACK_ARN | jq -r .Stacks[0].StackStatus)
                echo stack status of \${STACK_ARN} is \${STACK_STATUS}
              done
            - npm run -w apps/control-plane ampx -- generate outputs --branch $AWS_BRANCH --app-id $CONTROL_PLANE_APP_ID
            # 自分自身のリソースをデプロイする
            - npm run -w \${W} ampx -- pipeline-deploy --branch $AWS_BRANCH --app-id $AWS_APP_ID
",
        "CacheConfig": {
          "Type": "AMPLIFY_MANAGED_NO_COOKIES",
        },
        "CustomRules": [
          {
            "Source": "/<*>",
            "Status": "404-200",
            "Target": "/index.html",
          },
        ],
        "Description": "control plane",
        "EnableBranchAutoDeletion": false,
        "EnvironmentVariables": [
          {
            "Name": "AMPLIFY_DIFF_DEPLOY",
            "Value": "false",
          },
          {
            "Name": "AMPLIFY_MONOREPO_APP_ROOT",
            "Value": "apps/control-plane",
          },
        ],
        "IAMServiceRole": {
          "Fn::GetAtt": [
            "ControlPlaneServiceRole9996ABB7",
            "Arn",
          ],
        },
        "Name": "full-stack-silo-deploy-model-control-plane",
        "Platform": "WEB",
        "Repository": "https://github.com/horietakehiro/deep-dive-multi-tenant-saas-on-aws",
      },
      "Type": "AWS::Amplify::App",
    },
    "ControlPlaneBranch31C563C5": {
      "Properties": {
        "AppId": {
          "Fn::GetAtt": [
            "ControlPlaneApp8CAC7211",
            "AppId",
          ],
        },
        "BranchName": "main",
        "EnableAutoBuild": true,
        "Framework": "web",
        "Stage": "PRODUCTION",
      },
      "Type": "AWS::Amplify::Branch",
    },
    "ControlPlaneDomain7AAF1873": {
      "Properties": {
        "AppId": {
          "Fn::GetAtt": [
            "ControlPlaneApp8CAC7211",
            "AppId",
          ],
        },
        "AutoSubDomainCreationPatterns": [],
        "CertificateSettings": {
          "CertificateType": "CUSTOM",
          "CustomCertificateArn": "arn:aws:acm:Region:444455556666:certificate/certificate_ID",
        },
        "DomainName": "ht-burdock.com",
        "SubDomainSettings": [
          {
            "BranchName": "main",
            "Prefix": "control-plane",
          },
        ],
      },
      "Type": "AWS::Amplify::Domain",
    },
    "ControlPlaneServiceRole9996ABB7": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "amplify.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AmplifyBackendDeployFullAccess",
              ],
            ],
          },
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/ReadOnlyAccess",
              ],
            ],
          },
        ],
        "Policies": [
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": [
                    "logs:CreateLogStream",
                    "logs:PutLogEvents",
                    "logs:CreateLogGroup",
                    "logs:DescribeLogGroups",
                  ],
                  "Effect": "Allow",
                  "Resource": "*",
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "root",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
